### PR TITLE
[FLINK-30532] Add benchmark for DCT, SQLTransformer and StopWordsRemover

### PR DESCRIPTION
**Pull request by yunfengzhou-hub**
## What is the purpose of the change

This pull request mainly adds the benchmark configuration files for DCT, SQLTransformer and StopWordsRemover algorithm.

## Brief change log

- Adds the benchmark configuration files for DCT, SQLTransformer, and StopWordsRemover algorithm.
- Fixes the missing document of the DCT algorithm.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by yunfengzhou-hub** [comment]
Hi @jiangxin369, Could you please take a look at this PR?

File: flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/clustering/KMeansModelDataGenerator.java [hunk 1]
```diff
@@ -55,20 +61,36 @@ public Table[] getData(StreamTableEnvironment tEnv) {
...
+                        call(GenerateWeightsFunction.class, $("centroids")).as("weights"));
 
-        return new Table[] {tEnv.fromDataStream(modelDataStream)};
+        return new Table[] {modelDataTable};
     }
 
-    private static class GenerateKMeansModelDataFunction
-            implements MapFunction<DenseVector[], KMeansModelData> {
+    /**
+     * A scalar function that generate the weights vector for KMeansModelData from the centroids
```
Comment (jiangxin369) [inline flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/clustering/KMeansModelDataGenerator.java:77]:
```suggestion
     * A scalar function that generates the weights vector for KMeansModelData from the centroids
```

**Review by jiangxin369 — COMMENTED** [review]
Thanks for the PR, LGTM overall, only left one minor comment

File: flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/param/HasNumDistinctValues.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+package org.apache.flink.ml.benchmark.datagenerator.param;
+
+import org.apache.flink.ml.param.IntParam;
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.param.ParamValidators;
+import org.apache.flink.ml.param.WithParams;
+
+/** Interface for the benchmark num distinct values param. */
+public interface HasNumDistinctValues<T> extends WithParams<T> {
```
Comment (lindong28) [inline flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/param/HasNumDistinctValues.java:27]: Should we update `RandomStringGenerator` to use this class to reduce code duplication?

File: flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/clustering/KMeansModelDataGenerator.java [hunk 3]
```diff
@@ -55,20 +61,36 @@ public Table[] getData(StreamTableEnvironment tEnv) {
...
+        Table centroidsTable = vectorArrayGenerator.getData(tEnv)[0];
 
-        Table vectorArrayTable = vectorArrayGenerator.getData(tEnv)[0];
-        DataStream<KMeansModelData> modelDataStream =
-                tEnv.toDataStream(vectorArrayTable, DenseVector[].class)
-                        .map(new GenerateKMeansModelDataFunction());
+        Table modelDataTable =
+                centroidsTable.select(
+                        $("centroids"),
+                        call(GenerateWeightsFunction.class, $("centroids")).as("weights"));
```
Comment (lindong28) [inline flink-ml-benchmark/src/main/java/org/apache/flink/ml/benchmark/datagenerator/clustering/KMeansModelDataGenerator.java:71]: Can you explain why we need to make these changes and add names to those fields in the model data table?

Would it be better to keep the API (e.g. table format) consistent with that used by the production code (i.e. KMeansModel and KMeans)? For example, `KMeans#fit()` converts data stream to the model data table using `tEnv.fromDataStream(finalModelData)`.

Also, can we add a TODO/JIRA for creating nightly build to run all benchmarks nightly? We can use a small input number to verify that benchmark tools are broken.

**Review by lindong28 — COMMENTED** [review]
Thanks for the PR. Left some comments below.

File: flink-ml-benchmark/src/main/resources/dct-benchmark.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        ],
+        "seed": 2,
+        "numValues": 10000000,
+        "vectorDim": 100
+      }
+    },
+    "stage": {
+      "className": "org.apache.flink.ml.feature.dct.DCT",
+      "paramMap": {
+        "inputCol": "input",
```
Comment (lindong28) [inline flink-ml-benchmark/src/main/resources/dct-benchmark.json:35]: Would it be simpler to remove this line and use the default value? Same for outputCol.

Same for similar parameters used in other benchmark json files added in this PR.

**Comment by lindong28** [comment]
Thanks for the update. LGTM.

**Pull request closed without merging**